File: 备赛系统/B_历年真题实战/2024/B_单相功率分析仪/01_代码/Algorithm/dft_harmonic.c
/**
 * @file    dft_harmonic.c
 * @brief   单频 DFT × 10 谐波 + THD 实现
 */

#include "dft_harmonic.h"
#include <math.h>

#define TWO_PI  6.28318530718f

/* sin/cos LUT：每个谐波 bin 一份完整周期的 sin/cos
 *  尺寸：NUM_HARMONICS × DFT_N × 2(sin + cos) × 4 字节
 *      = 10 × 1000 × 8 = 80 KB → MSP430 SRAM 不够
 *  对策：只存"基波"一份 LUT，谐波通过下标乘法在线索引。
 *      sin(2π · h · n / N) = sin( 2π · (h × n) mod N / N )
 *      用基波 LUT 直接查表，节省 90% 内存。
 */
static float s_sin_lut[DFT_N];
static float s_cos_lut[DFT_N];
static uint8_t s_lut_ready = 0;

void DFTHarm_Init(void)
{
    if (s_lut_ready) return;
    for (uint16_t k = 0; k < DFT_N; k++) {
        float angle = TWO_PI * (float)k / (float)DFT_N;
        s_sin_lut[k] = sinf(angle);
        s_cos_lut[k] = cosf(angle);
    }
    s_lut_ready = 1u;
}

/* 单频 DFT：计算 X(bin) 的幅值 */
static float dft_bin_magnitude(const float *x, uint16_t n, uint16_t bin)
{
    /* 用 double 累积防止大数据集精度损失 */
    double re = 0.0;
    double im = 0.0;

    for (uint16_t k = 0; k < n; k++) {
        /* (bin × k) mod N → LUT 索引 */
        uint32_t lut_idx = ((uint32_t)bin * (uint32_t)k) % (uint32_t)n;
        re += (double)x[k] * (double)s_cos_lut[lut_idx];
        im -= (double)x[k] * (double)s_sin_lut[lut_idx];
    }

    /* 单边幅值：|X| × 2/N */
    double mag = sqrt(re * re + im * im) * 2.0 / (double)n;
    return (float)mag;
}

void DFTHarm_Compute(const float *i_samples, uint16_t n, HarmonicResult *out)
{
    if (!i_samples || !out || n != DFT_N) {
        if (out) {
            for (uint8_t h = 0; h <= NUM_HARMONICS; h++) {
                out->magnitude[h] = 0.0f;
                out->rms[h] = 0.0f;
                out->h_norm[h] = 0.0f;
            }
            out->thd_percent = 0.0f;
        }
        return;
    }

    if (!s_lut_ready) DFTHarm_Init();

    /* ---- 提取各次谐波 ---- */
    for (uint8_t h = 1; h <= NUM_HARMONICS; h++) {
        uint16_t bin = (uint16_t)FUND_BIN * (uint16_t)h;
        if (bin >= (uint16_t)(DFT_N / 2u)) {
            out->magnitude[h] = 0.0f;
            out->rms[h] = 0.0f;
            continue;
        }
        out->magnitude[h] = dft_bin_magnitude(i_samples, n, bin);
        out->rms[h] = out->magnitude[h] * 0.7071068f;   /* 峰值 → RMS */
    }

    /* ---- 归一化 + THD ---- */
    out->h_norm[1] = 1.0f;
    if (out->rms[1] > 1e-6f) {
        double sum_harm_sq = 0.0;
        for (uint8_t h = 2; h <= NUM_HARMONICS; h++) {
            out->h_norm[h] = out->rms[h] / out->rms[1];
            sum_harm_sq += (double)out->rms[h] * (double)out->rms[h];
        }
        out->thd_percent = (float)(sqrt(sum_harm_sq) / (double)out->rms[1] * 100.0);
    } else {
        for (uint8_t h = 2; h <= NUM_HARMONICS; h++) out->h_norm[h] = 0.0f;
        out->thd_percent = 0.0f;
    }
}

File: 备赛系统/C_通用能力/算法层PC验证方法论/snippets/dft_harmonic.py
"""
dft_harmonic.py
===============

单频 DFT × N 谐波 + THD 计算 Python 参考实现。
（与 2024 B 的 dft_harmonic.c + 2021 A 的 thd_measure.c 等价）

核心思路：
  - 不做完整 FFT，只算需要的 N 个 bin（基波 + 谐波）
  - 共用基波 sin/cos LUT，谐波通过 (h × n) mod N 索引
  - 整周期采样下用矩形窗（无频谱泄漏，精度最高）

用法：
    from dft_harmonic import DFTHarmonic
    
    dft = DFTHarmonic(n_samples=1000, fund_bin=5, num_harmonics=10)
    result = dft.compute(samples)
    print(result.thd_percent, result.h_norm)
"""

import math
from dataclasses import dataclass, field
from typing import List


@dataclass
class HarmonicResult:
    magnitude: List[float] = field(default_factory=list)  # [0..N]
    rms: List[float]       = field(default_factory=list)
    h_norm: List[float]    = field(default_factory=list)
    thd_percent: float = 0.0


class DFTHarmonic:
    """整周期采样下的单频 DFT。
    
    Args:
        n_samples: DFT 长度（必须等于 整数周期 × 每周期点数）
        fund_bin:  基波在 DFT 中的 bin 索引（必须是整数）
        num_harmonics: 提取的谐波次数（含基波）
    """

    def __init__(self, n_samples: int, fund_bin: int, num_harmonics: int = 10):
        self.n = n_samples
        self.fund_bin = fund_bin
        self.num = num_harmonics
        # 预计算 sin/cos LUT
        self.sin_lut = [math.sin(2 * math.pi * k / n_samples) for k in range(n_samples)]
        self.cos_lut = [math.cos(2 * math.pi * k / n_samples) for k in range(n_samples)]

    def _bin_magnitude(self, x: List[float], bin_idx: int) -> float:
        re = im = 0.0
        for k in range(self.n):
            lut_idx = (bin_idx * k) % self.n
            re += x[k] * self.cos_lut[lut_idx]
            im -= x[k] * self.sin_lut[lut_idx]
        return math.sqrt(re * re + im * im) * 2.0 / self.n

    def compute(self, samples: List[float]) -> HarmonicResult:
        if len(samples) != self.n:
            raise ValueError(f"样本数 {len(samples)} != DFT 长度 {self.n}")

        result = HarmonicResult(
            magnitude=[0.0] * (self.num + 1),
            rms=[0.0] * (self.num + 1),
            h_norm=[0.0] * (self.num + 1),
        )

        # 提取各次谐波
        for h in range(1, self.num + 1):
            bin_idx = self.fund_bin * h
            if bin_idx >= self.n // 2:
                continue
            mag = self._bin_magnitude(samples, bin_idx)
            result.magnitude[h] = mag
            result.rms[h] = mag / math.sqrt(2.0)

        # 归一化 + THD
        result.h_norm[1] = 1.0
        if result.rms[1] > 1e-9:
            sum_sq = 0.0
            for h in range(2, self.num + 1):
                result.h_norm[h] = result.rms[h] / result.rms[1]
                sum_sq += result.rms[h] ** 2
            result.thd_percent = math.sqrt(sum_sq) / result.rms[1] * 100.0

        return result


def _example():
    """1kHz 基波 + H3=20% + H5=10% → 期望 THD = √(0.2² + 0.1²) × 100 = 22.36%"""
    fs = 20000.0    # 20 kHz（H5 = 5kHz，留充足奈奎斯特余量）
    freq = 1000.0
    # 1kHz @ 20kHz 采样，1000 点 → 50 个完整周期，fund_bin = 1000 × 1000 / 20000 = 50
    n_samples = 1000
    fund_bin = 50

    samples = [
        math.sin(2 * math.pi * freq * k / fs)
        + 0.2 * math.sin(2 * math.pi * 3 * freq * k / fs)
        + 0.1 * math.sin(2 * math.pi * 5 * freq * k / fs)
        for k in range(n_samples)
    ]

    dft = DFTHarmonic(n_samples=n_samples, fund_bin=fund_bin, num_harmonics=10)
    r = dft.compute(samples)

    expected_thd = math.sqrt(0.2 ** 2 + 0.1 ** 2) * 100
    print(f"基波幅值      : {r.magnitude[1]:.6f} (期望 1.0)")
    print(f"H3 / H1       : {r.h_norm[3]:.6f} (期望 0.2)")
    print(f"H5 / H1       : {r.h_norm[5]:.6f} (期望 0.1)")
    print(f"THD           : {r.thd_percent:.6f}% (期望 {expected_thd:.6f}%)")
    err = abs(r.thd_percent - expected_thd)
    print(f"误差          : {err:.6f}% {'✓' if err < 1e-4 else '✗'}")


if __name__ == "__main__":
    _example()
